smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,464 @@
+
+        Parameters
+        ----------
+        challengers : typing.List[Configuration]
+            promising configurations
+        incumbent : Configuration
+            best configuration so far
+        run_history : RunHistory
+            stores all runs we ran so far
+        aggregate_func: typing.Callable
+            aggregate error across instances
+        time_bound : float, optional (default=2 ** 31 - 1)
+            time in [sec] available to perform intensify
+        log_traj: bool
+            whether to log changes of incumbents in trajectory
+
+        Returns
+        -------
+        typing.Tuple[Configuration, float]
+        """
+        self.start_time = time.time()
+        self._ta_time = 0
+        self._chall_indx = 0
+        self._num_run = 0
+        first_run = True
+
+        # challengers can be repeated only if optimizing across multiple seeds or changing instance orders every run
+        if not (self.n_seeds > 1 or self.instance_order == 'shuffle'):
+            used_configs = set(run_history.get_all_configs())
+            chal_generator = (c for c in challengers if c not in used_configs)
+        else:
+            # converting to generator for consistency
+            chal_generator = (c for c in challengers)
+        # select first 'n' challengers
+        curr_challengers = list(islice(chal_generator, self.num_initial_challengers))
+
+        # randomize instances per successive halving run, if user specifies
+        all_instances = self.instances
+        if self.instance_order == 'shuffle':
+            self.rs.shuffle(all_instances)
+
+        # calculating the incumbent's performance for adaptive capping
+        #   - this check is required because there is no incumbent performance
+        #     for the first ever 'intensify' run (from initial design)
+        if incumbent is not None:
+            inc_runs = run_history.get_runs_for_config(incumbent)
+            inc_sum_cost = sum_cost(config=incumbent, instance_seed_budget_keys=inc_runs, run_history=run_history)
+        else:
+            inc_sum_cost = np.inf
+
+        self.logger.debug('---' * 40)
+        self.logger.debug('Successive Halving run begins. Budgets: %s' % self.budgets)
+
+        # run intensification till budget is max
+        for i, curr_budget in enumerate(self.budgets):
+
+            self.logger.info('Running with budget [%.2f / %d] with %d challengers' %
+                             (curr_budget, self.max_budget, len(curr_challengers)))
+            # selecting instance subset for this budget, depending on the kind of budget
+            prev_budget = self.budgets[i - 1] if i > 0 else 0
+            available_insts = all_instances[int(prev_budget):int(curr_budget)] if not self.cutoff_as_budget \
+                else all_instances
+
+            # determine 'k' for the next iteration - at least 1
+            next_n_chal = max(1, int(np.round(len(curr_challengers) / self.eta)))
```

Comment:
Could you please use `np.floor` instead of `np.round` to make the code closer to the pseudo-code in the paper?